libgalois/src/analytics/connected_components/connected_components.cpp

```diff
@@ -66,15 +67,16 @@ struct ConnectedComponentsSerialAlgo {
 
   void Initialize(Graph* graph) {
     katana::do_all(katana::iterate(*graph), [&](const GNode& node) {
-      graph->GetData<NodeComponent>(node) = new ConnectedComponentsNode();
+      auto& sdata = graph->GetData<NodeComponent>(node);
+      static_assert(sizeof(ConnectedComponentsNode) == sizeof(uint64_t));
+      new (&sdata) ConnectedComponentsNode();
```

Comment:
This looks like it's constructing a `ConnectedComponentsNode` in a memory location of type `ConnectedComponentsNode*`. I think this works because placement new takes `void*` so it will coerce the pointer. And we get away with it because `ConnectedComponentsNode` is the same size as a pointer. However this seems massively unsafe and bad. And could cause problems if there are alignment requirements on atomic values that don't exist for normal pointers. Even if your assert checks for size issues.

Can we change `ComponentType` to be `ConnectedComponentsNode`? Or maybe just change `NodeComponent` to use a view type of `ConnectedComponentsNode`?

Reply:
Also doesn't having the type wrong make it easy to accidentally perform a non-atomic access even though accesses may need to be atomic?